### PR TITLE
Typst CSS and Raw HTML documentation

### PR DESCRIPTION
**Pull request by gordonwoodhull**
This is getting closer to ready...

fixes https://github.com/quarto-dev/quarto-cli/issues/9881

### CODE REVIEW

**Comment by gordonwoodhull** [comment]
This also contains initial documentation of features that are only available through Raw HTML:
* `colspan`
* `rowspan`
* `text-align`
* `vertical-align`

**Comment by gordonwoodhull** [comment]
Open questions for this PR:
- [ ] Is this a good place / way to describe the "Raw HTML features"? (Really HTML/CSS features that haven't been documented before). I added these to the ff matrix - looks like colspan/rowspan have universal support but cell alignment is spotty:
image
- [ ] Where to put the Typst table styling examples / how to screenshot them, as discussed. I removed them because they were crashing. Will follow up in the next comment with links and screenshots of the examples I'd like to use, which show a nice variety of features.
- [ ] I documented `css-property-processing` inside the advanced Typst CSS doc, still need to implement this (quarto-dev/quarto-cli#9885) 
- [ ] The matrix of element/property support in Typst CSS is more sparse than I thought. Does this look bad? A lot more is possible but I focused only on the table & pre use-cases.

Apologies if I am not following proper syntax or conventions for the docs, this is my first PR. All feedback is welcome!

**Comment by gordonwoodhull** [comment]
The examples I'd like to use are all smoke tests currently. I have borrowed them from blogs and documentation. 

(Side question: is there a tag for `example-source`?)

## flextable
image

https://github.com/quarto-dev/quarto-cli/blob/main/tests/docs/typst/tables/flextable/viridis-cells.qmd

## gt
image

https://github.com/quarto-dev/quarto-cli/blob/main/tests/docs/typst/tables/gt/gtcars.qmd

## pandas
image

https://github.com/quarto-dev/quarto-cli/blob/main/tests/docs/typst/tables/pandas/data-cell-css-classes.qmd

Another possibility would be to code the same example using all three packages, but that would be a good amount of work.

**Comment by cwickham** [comment]
A few answers, that I can provide:

> Is this a good place / way to describe the "Raw HTML features"? 

I think it is the best spot we currently have. 

> Where to put the Typst table styling examples & how to screenshot them, as discussed Monday. I've put in placeholder images for now.

It sounded like @cderv had some ideas for this. Let's defer to him. 

> Another possibility would be to code the same example using all three packages, but that would be a good amount of work.

I actually like this idea, but agree it could be a lot of work. Let me play with it for a bit and get back to you.

FYI `docs/authoring/tables.qmd` has executed code cells in it, so it requires an update to `_freeze/`. You can locally render and commit the update, or just ping me to do it, since I have everything set up. (I'll do it once now, just so I can look at the preview).

**Review by cwickham — COMMENTED** [review]
Looking really good Gordon! A number of little tweaks and just a couple of bigger ones.

**Comment by cwickham** [comment]
Oh, one other thing. Is it worth explaining how someone might determine if a package/library emits HTML? Maybe that's something we could do in the blog post instead...

File: docs/output-formats/typst.qmd [hunk 1]
```diff
@@ -200,6 +200,47 @@ In the case of glaciers, fluid dynamics principles can be used to understand how
...
+![](images/typst-table-gt.png)
+
+[gt cars example source](examples/gt-cars.qmd)
+
+#### Python - pandas
+![](images/typst-table-pandas.png)
+
+[pandas confusion matrix example source](examples/pandas-confusion-matrix.qmd)
+
+:::
```
[the remark this answers is not in the document]
Comment (gordonwoodhull) [reply docs/output-formats/typst.qmd:239]: Okay, let's do a simple example that only shows background color. That's enough to get people started.

How about using a temperature heatmap like this example:
https://themockup.blog/static/resources/gt-cookbook.html#multiple-columns

It's about a dozen lines of code, and should be pretty easy to translate to Pandas.

[the remark this answers is not in the document]
Comment (cwickham) [reply docs/output-formats/typst.qmd:239]: Yeah, let's do something like that. But probably a little smaller with hardcoded data. Maybe:
```
temps <- data.frame(
  year = c(1920:1924),
  Jan = c(40.6, 44.2, 37.5, 41.8, 39.3), 
  Jun = c(58.5, 58.7, 57.8, 52.7, 57.7)
)

temps %>% 
  gt() %>% 
  data_color(
    columns = c(-year),
    colors = scales::col_numeric(
      c("#D33F6A", "#E99A2C", "#E2E6BD"), 
      domain = c(30, 70))
  )
```

**Comment by gordonwoodhull** [comment]
> Oh, one other thing. Is it worth explaining how someone might determine if a package/library emits HTML? Maybe that's something we could do in the blog post instead...

Yeah, I think the advanced doc should have a troubleshooting guide. The first step is using `keep-md:true` and looking for an HTML raw block

````
```{=html}
````

with a `<table>` in it.

We can include that hint in the block post, too. Good idea.

All excellent feedback, thanks!

File: docs/authoring/tables.qmd [hunk 2]
```diff
@@ -551,6 +551,44 @@ Which renders as follows:
...
+  <td style="text-align: left">C2</td>
+ </tr>
+ <tr>
+  <th>3</th>
+  <td style="vertical-align: top">B3</td>
+  <td style="text-align: right">C3</td>
+ </tr>
+</table>
+```
+````
```
[the remark this answers is not in the document]
Comment (gordonwoodhull) [reply docs/authoring/tables.qmd:589]: Hmm, this is the example that I used for the feature-format matrix. It isn't pretty

image

but you can see whether the spans and aligns are working in each format.

Two separate examples, for `colspan`/`rowspan` and for `text-align`/`vertical-align`, would each be shorter, but would be longer in total. Would that be better?

Is there a way to "echo" Quarto markdown? Output the block as quoted qmd, and then output it unquoted?

File: docs/output-formats/typst.qmd (hunk 1 above)
[the remark this answers is not in the document]
Comment (gordonwoodhull) [reply docs/output-formats/typst.qmd:239]: Nice

File: docs/authoring/tables.qmd (hunk 2 above)
[the remark this answers is not in the document]
Comment (cwickham) [reply docs/authoring/tables.qmd:589]: Yeah, an include should work for that. E.g. put the table source in `_table-example.md`. Then you should be able to use it like:
````
{{< include _table-example.md >}}

```html
{{< include _table-example.md >}}
```

````

[the remark this answers is not in the document]
Comment (cwickham) [reply docs/authoring/tables.qmd:589]: Oh wait, I don't know how that will play with the raw block...I try it.

[the remark this answers is not in the document]
Comment (cwickham) [reply docs/authoring/tables.qmd:589]: It worked fine (at least in isolation):

`````
{{< include _table-example.md >}}

````html
{{< include _table-example.md >}}
````
`````

Where `_table-example.md` contains:
````
```{=html}
<table>
 ...
</table>
```
````

File: docs/advanced/typst/typst-css.qmd [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+It is tuned for two use-cases:
+
+- HTML Tables styled by table packages
+- HTML `<pre>` tags representing console output with spans decorated with colors, e.g. from ANSI.
+
+This document is oriented to
+* package developers who want to produce `table` or `pre` output which can be translated to Typst
+* extension developers who want to improve the translation of CSS properties, or add other CSS properties
+
+It describes how Typst CSS works, how to enable and disable this feature and `pre` tag processing
```
[the remark this answers is not in the document]
Comment (gordonwoodhull) [reply docs/advanced/typst/typst-css.qmd:16]: This is the start of a longer list (which I must have accidentally truncated) and within it are two features (Typst CSS itself and `html-pre-tag-processing`) which can be enabled or disabled. 

I'm not sure if we need to list everything, given there is a table of contents to the right. So I'm completing the list but maybe we should remove it instead.

File: docs/authoring/tables.qmd [hunk 4]
```diff
@@ -551,6 +551,44 @@ Which renders as follows:
...
+
+These features are available in PDF, Typst, and Docx in addition to HTML.
```
[the remark this answers is not in the document]
Comment (gordonwoodhull) [reply docs/authoring/tables.qmd:591]: This doesn't make sense, removing it. 

I think this section will be clearer with colspan/rowspan separated from align.

**Comment by gordonwoodhull** [comment]
I've committed all the minor copy edits as well as new temperature examples in gt and Pandas.

This PR also contains a new tool to render a Typst document, suppressing non-essential output, snapshotting to PNG, and trimming.

One minor glitch: I am not getting syntax highlighting on the gt example. I'm trying to do

`````
````r
```{{r}}
{{< include examples/gt-temps.R >}}
```
````
`````

but it seems the usual trick of marking the code as R instead of markdown for syntax highlighting does not work because of the backticks. Same trick does work for Python.

I still need to 

- [ ] Split the colspan/rowspan and align examples and surrounding text.
- [ ] Integrate the snapshot tool - no luck with pre-render scripts so far.
- Test modifying the Typst CSS translation with user-defined Lua filters.
  - [ ] Filter override with a pre-quarto filter removing CSS properties.
  - [ ] Handling a new property post-render.

**Comment by cwickham** [comment]
Hmm...I feel like the R syntax highlighting is something I've struggled to get working before too (outside of an `include`). I'll ask in Slack.

File: docs/advanced/typst/typst-css.qmd (hunk 3 above)
[the remark this answers is not in the document]
Comment (cwickham) [reply docs/advanced/typst/typst-css.qmd]: I quite like the "Overview" to briefly introduce the sections, so I'd vote for completing the list.

**Comment by gordonwoodhull** [comment]
Very strange, I force-pushed this branch and the PR closed itself. It says it was merged but thankfully it is not. Will start a replacement PR.

**Comment by gordonwoodhull** [comment]
reopening

**Pull request closed without merging**